Live Mempool Page

RogueElement opened the pull request:
For issues #230 and #129 
Todo:
- [x] Mini live transaction viewer on front page
- ~~Extra mempool page for votes and revocations~~
Mempool page is now similar to block page

raedah commented:
- Add 'Type' field on the Latest Transactions to show Vote, Ticket, Revoke, Transaction
 - On mempool page, should say Mempool as a top header. 
 - On mempool page, swap the boxes so Tickets list is first on the left and then Tickets next.
 - Instead of "More transactions and tickets..." could say "More mempool..."  
 - Number of rows in Latest transactions is growing dynamically. Should be statically fixed, make it 5.
 - Make the grey header bar not be double tall that would look better.
 - Get the hash row to the size so it wraps to 2 lines instead of leaving white space.
 - Horizontally align latest transactions with latest blocks. Put transactions on the top, blocks on the bottom.
 - Utilize more space on the width of that home page since we are showing more there. 
 - Make latest blocks 5 rows too.

RogueElement commented:
So far I've:
- added a type field,
- changed "More transactions and tickets..." to "More mempool..."
- made number of latest transaction rows static and put it on top of latest blocks
- made latest blocks 5 rows

I don't know how to do the rest cause they require styling.

raedah commented:
This is working well, up until these errors occur and the server stops responding,
2018-02-15 06:33:00.426 [DBG] EXPR: Registering new websocket client
2018-02-15 06:35:04.359 [WRN] DATD: expNewTxChan buffer full!
2018-02-15 06:36:19.813 [WRN] EXPR: websocket client receive error: EOF
2018-02-15 06:36:19.813 [WRN] EXPR: unknown client

Other things we can add to this..
TOTAL SENT  (sum dcr out of all tx in mempool)
SIZE  (sum size of all tx in the mempool)
BEST BLOCK  (block # and time ago)
For consistancy, put these fields in roughly the same spot they are on the block page.

To be consistent with the block page, when no transactions of a type are in the mempool, we would not show the headers and show the "No tickets in the mempool." msg, sub for whatever type it is.

Also for consistency, use Total DCR instead of Total Out, and we should standardize on either Transactions ID or Hash.

For the Votes we can display the same additional columns of data which are on the block page.

Move the 'Time in Mempool' column to be the last column on the right.

Use a capital M in the HTML page title.

From the best block page, NEXT BLOCK could link to mempool, and for the text say 'mempool' instead of dash. https://github.com/dcrdata/dcrdata/blob/master/views/block.tmpl#L57

raedah commented:
tACK

chappjc commented:
Sweet.  Will review

chappjc commented on db/dcrsqlite/apisource.go:
```diff
@@ -1235,3 +1235,54 @@ func (db *wiredDB) CountUnconfirmedTransactions(address string, maxUnconfirmedPo
 	return
 }
+
+func (db *wiredDB) GetMempool() []explorer.MempoolTx {
```
Please doc.

chappjc commented on explorer/mempool.go:
```diff
@@ -0,0 +1,183 @@
+package explorer
```
standard copyright

chappjc commented on explorer/explorertypes.go:
```diff
@@ -277,3 +295,12 @@ type TicketPoolInfo struct {
 	PercentTarget float64 `json:"percent_target"`
 }
+
+type MempoolTx struct {
```
doc

chappjc commented:
Please rebase and resolve conflicts.  Just checked... nothing tricky.

RogueElement commented:
Rebased and resolved.

raedah commented:
Are all known instances of the sockets hanging fixed?

raedah commented:
Total Sent is slightly out of alignment to the right.

raedah commented:
Note, we can also add links to the mempool api, though doesnt need to be in this PR.

chappjc commented on ntfnhandlers.go:
```diff
@@ -225,8 +240,56 @@ func makeNodeNtfnHandlers(cfg *config) (*rpcclient.NotificationHandlers, *collec
+		// for the explorer mempool monitor to avoid an extra call to dcrd for
+		// the tx details
+		OnTxAcceptedVerbose: func(txDetails *dcrjson.TxRawResult) {
+			msgTx := txhelpers.MsgTxFromHex(txDetails.Hex)
```
In general, we want to do as little as possible in these On* handlers since while they are being executed, the RPC client is blocked.  This is why I try to send the bare minimum on a channel to the actual handlers downstream, with a hefty buffer.

So sending the msgTx might be a good idea, then leave the function calls (IsSSGen, SSGenVoteChoices, etc.) to whatever function reads the channel.

chappjc replied:
If we're still seeing issues with hanging and backed up buffered channels, we should look carefully at these handlers.

chappjc commented on db/dcrsqlite/apisource.go:
```diff
@@ -1262,3 +1262,56 @@ func (db *wiredDB) CountUnconfirmedTransactions(address string, maxUnconfirmedPo
 	return
 }
+
+// GetMepool gets all transactions from the mempool for explorer
```
Sometime in the future we need to go through and remove all "Get" prefixes where it's absolutely not necessary.  This is good for now, just a heads up.

chappjc commented on db/dcrsqlite/apisource.go:
```diff
@@ -1262,3 +1262,56 @@ func (db *wiredDB) CountUnconfirmedTransactions(address string, maxUnconfirmedPo
+		for _, v := range rawtx.Vout {
+			total += v.Value
+		}
+		msgTx := txhelpers.MsgTxFromHex(hex)
```
Need to check if its nil, or better, refactor MsgTxFromHex to return its error.

chappjc commented:
I suspect duplicate txns are being sent to our handlers via OnTxAcceptedVerbose.  Perhaps we should check for duplicate votes, or look into speeding up the new tx handers somehow.

chappjc commented:
please rebase and squash at least partially.

RogueElement commented:
Rebased and squashed

chappjc commented on db/dcrsqlite/apisource.go:
```diff
@@ -973,8 +972,8 @@ func (db *wiredDB) GetExplorerBlock(hash string) *explorer.BlockInfo {
-		msgTx := txhelpers.MsgTxFromHex(tx.Hex)
-		if msgTx == nil {
+		msgTx, err := txhelpers.MsgTxFromHex(tx.Hex)
+		if err != nil {
```
please include `err` in the log message below since we're not returning it from this functions

chappjc commented on db/dcrsqlite/apisource.go:
```diff
@@ -1053,8 +1056,8 @@ func (db *wiredDB) GetExplorerTx(txid string) *explorer.TxInfo {
-	if msgTx == nil {
+	msgTx, err := txhelpers.MsgTxFromHex(txraw.Hex)
+	if err != nil {
 		log.Errorf("Cannot create MsgTx for tx %v", txhash)
```
log the `err` here too please

chappjc commented on explorer/explorer.go:
```diff
@@ -427,6 +438,15 @@ func New(dataSource explorerDataSourceLite, primaryDataSource explorerDataSource
+		exp.templateFiles["extras"],
+	)
+	if err != nil {
+		log.Errorf("Unable to create new html template: %v", err)
```
Please use `noTemplateError(err)` now.  (This changed when you rebased)

chappjc commented on explorer/explorerroutes.go:
```diff
@@ -16,10 +16,10 @@ import (
 
 	exp.NewBlockDataMtx.Lock()
-	exp.MempoolData.RLock()
+	exp.MempoolData.Lock()
```
Why can't this be `RLock`?

RogueElement replied:
That slipped by, I was testing something. Changed back now.

chappjc commented on explorer/explorerroutes.go:
```diff
@@ -135,6 +135,28 @@ func (exp *explorerUI) Block(w http.ResponseWriter, r *http.Request) {
 
+// Mempool is the page handler for the "/mempool" path
+func (exp *explorerUI) Mempool(w http.ResponseWriter, r *http.Request) {
+	exp.MempoolData.Lock()
```
Why can't this be `RLock`?

chappjc commented on explorer/mempool.go:
```diff
@@ -0,0 +1,228 @@
+			time.Since(start))
+	}(time.Now())
+
+	exp.MempoolData.Lock()
```
I'm thinking we should move this closer to where we actually need it locked.  Even after the exp.NewBlockDataMtx.Lock() block as far as I can tell

chappjc commented on explorer/mempool.go:
```diff
@@ -0,0 +1,228 @@
+type byType []MempoolTx
+
+func (txs byType) Less(i, j int) bool {
+	weights := map[string]int{
```
this gets constructed every time `Less` is run.  Lets make it a global or deal with it some other way.

chappjc replied:
for instance, we could ditch the `sort.Interface` approach used her, and use[ `sort.Slice`](https://golang.org/pkg/sort/#Slice) where the specified closure (`less`) would capture the map if it were defined as a meber of `explorerUI`.  Just a though

RogueElement replied:
Removed `ByType` entirely, it is no longer needed since the transactions are already sorted into their respective slices but I'll keep in mind the `sort.Slice` approach.

chappjc commented on ntfnhandlers.go:
```diff
@@ -34,9 +36,13 @@ func registerNodeNtfnHandlers(dcrdClient *rpcclient.Client) *ContextualError {
-	if err = dcrdClient.NotifyNewTransactions(false); err != nil {
-		return newContextualError("new transaction "+
-			"notification registration failed", err)
+	// if err = dcrdClient.NotifyNewTransactions(false); err != nil {
```
please remove this bit of old code

chappjc commented on ntfnhandlers.go:
```diff
@@ -225,8 +238,33 @@ func makeNodeNtfnHandlers(cfg *config) (*rpcclient.NotificationHandlers, *collec
+		// the tx details
+		OnTxAcceptedVerbose: func(txDetails *dcrjson.TxRawResult) {
+			select {
+			case ntfnChans.expNewTxChan <- &explorer.NewMempoolTx{
```
👍

chappjc commented:
Looking pretty darn good now.  I haven't reviewed any JS or HTML yet though.

raedah commented:
tACK

chappjc commented:
This tests really well.  Nice any clean implementation.
Only thing is that this doesn't really provide any useful information, just confusion:

VOTES IN MEMPOOL | 28
-- | --

I think we need to filter out votes that are not for the tip block, and combine votes spending the same ticket, and only show a count from 0-5.

chappjc commented on explorer/websocket.go:
```diff
@@ -121,21 +124,28 @@ func (wsh *WebsocketHub) run() {
 				log.Infof("Signaling mempool update to %d websocket clients.", len(wsh.clients))
+			case sigNewTx:
+				newtx = <-wsh.NewTxChan
+				log.Infof("Signaling new tx %s to %d websocket clients.", newtx.Hash, len(wsh.clients))
```
This happens too often for `Info` level logging.  Please make Debug

chappjc commented:
In the future, I'd like to consider sending txns in batch to websocket clients.  So after say a 5 second delay, send whatever new txns we have, or if there are >N txns waiting, send them regardless of the delay.

chappjc commented:
Now running on https://beta.dcrdata.org

chappjc commented:
The votes table wraps the tx hash, but it could be on one line with small tweaks.  Window is full size here:

![image](https://user-images.githubusercontent.com/9373513/36992053-88d30f04-205e-11e8-9063-438f65fa6bfb.png)

gozart1 commented:
I've working on a flexbox based layout for these tables that allows for some more fine grained control over table row dimensions and truncates the hash to the space available. Also planning to add some animation to delineate new items being added.

The PR is still WIP, so maybe best to merge this without worrying too much about getting layout perfect and I'll make a separate PR for my frontend stuff.
https://github.com/gozart1/dcrdata-explorer/commit/ff1bc315ebd3acbf5dffe5a90925c442488b2b7f
Should have it ready in a few days.

chappjc commented:
ok
